File: legacy/src/retraction_pollution/openalex.py
from __future__ import annotations

import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass
from typing import Any

from .util import (
    clean_doi,
    compact_openalex_id,
    doi_url,
    full_openalex_id,
    json_dumps,
    text_or_none,
)

OPENALEX_API = "https://api.openalex.org"

WORK_SELECT = ",".join(
    [
        "id",
        "doi",
        "display_name",
        "title",
        "publication_date",
        "publication_year",
        "type",
        "type_crossref",
        "is_retracted",
        "cited_by_count",
        "referenced_works",
        "primary_location",
        "primary_topic",
        "topics",
    ]
)

RESOLUTION_SELECT = ",".join(
    [
        "id",
        "doi",
        "display_name",
        "title",
        "publication_date",
        "publication_year",
        "type",
        "is_retracted",
        "cited_by_count",
        "referenced_works",
        "primary_location",
        "primary_topic",
        "topics",
        "authorships",
    ]
)


class OpenAlexError(RuntimeError):
    def __init__(
        self,
        message: str,
        *,
        status_code: int | None = None,
        response_text: str | None = None,
    ):
        super().__init__(message)
        self.status_code = status_code
        self.response_text = response_text


class OpenAlexRateLimitError(OpenAlexError):
    pass


@dataclass
class OpenAlexPage:
    results: list[dict[str, Any]]
    next_cursor: str | None
    count: int | None


class OpenAlexClient:
    def __init__(
        self,
        api_key: str | None = None,
        email: str | None = None,
        base_url: str = OPENALEX_API,
        retries: int = 6,
        request_delay: float = 0.35,
        rate_limit_sleep: float = 60.0,
    ):
        self.api_key = api_key
        self.email = email
        self.base_url = base_url.rstrip("/")
        self.retries = retries
        self.request_delay = request_delay
        self.rate_limit_sleep = rate_limit_sleep

    def _request_json(self, path: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
        query: dict[str, Any] = dict(params or {})
        if self.api_key:
            query["api_key"] = self.api_key
        if self.email:
            query["mailto"] = self.email
        url = f"{self.base_url}{path}"
        if query:
            url = f"{url}?{urllib.parse.urlencode(query, doseq=True)}"

        last_error: Exception | None = None
        for attempt in range(self.retries):
            if attempt or self.request_delay:
                time.sleep(self.request_delay if attempt == 0 else min(60.0, 2.0**attempt))
            request = urllib.request.Request(
                url,
                headers={
                    "Accept": "application/json",
                    "User-Agent": "RetractionPollution/0.1 (mailto optional; OpenAlex research)",
                },
            )
            try:
                with urllib.request.urlopen(request, timeout=60) as response:
                    import json

                    return json.loads(response.read().decode("utf-8"))
            except urllib.error.HTTPError as exc:
                last_error = exc
                if exc.code == 404:
                    return {}
                if exc.code == 429:
                    retry_after = exc.headers.get("Retry-After")
                    time.sleep(_retry_after_seconds(retry_after, self.rate_limit_sleep, attempt))
                    continue
                if 500 <= exc.code < 600:
                    continue
                response_text = _read_http_error_body(exc)
                raise OpenAlexError(
                    f"OpenAlex HTTP {exc.code} for {path}: {response_text[:300]}",
                    status_code=exc.code,
                    response_text=response_text,
                ) from exc
            except urllib.error.URLError as exc:
                last_error = exc
                continue
        if isinstance(last_error, urllib.error.HTTPError) and last_error.code == 429:
            raise OpenAlexRateLimitError(
                "OpenAlex rate limit persisted after retries. The crawl state is saved; "
                "rerun the same command later to continue.",
                status_code=429,
                response_text=_read_http_error_body(last_error),
            ) from last_error
        raise OpenAlexError(f"OpenAlex request failed after {self.retries} attempts: {path}") from (
            last_error
        )

    def get_works_by_dois(self, dois: list[str]) -> list[dict[str, Any]]:
        values = list(dict.fromkeys(_openalex_doi_url(doi) for doi in dois))
        values = [value for value in values if value]
        if not values:
            return []
        if len(values) > 100:
            works: list[dict[str, Any]] = []
            for start in range(0, len(values), 100):
                works.extend(self._get_works_by_doi_values(values[start : start + 100]))
            return works
        return self._get_works_by_doi_values(values)

    def rate_limit_status(self) -> dict[str, Any]:
        return self._request_json("/rate-limit")

    def _get_works_by_doi_values(self, values: list[str]) -> list[dict[str, Any]]:
        if not values:
            return []
        try:
            data = self._request_json(
                "/works",
                {
                    "filter": "doi:" + "|".join(values),
                    "per-page": max(1, min(100, len(values))),
                    "select": RESOLUTION_SELECT,
                },
            )
            return data.get("results") or []
        except OpenAlexError as exc:
            if exc.status_code != 400:
                raise
            if len(values) == 1:
                return []
            midpoint = len(values) // 2
            return self._get_works_by_doi_values(
                values[:midpoint]
            ) + self._get_works_by_doi_values(values[midpoint:])

    def get_work_by_pmid(self, pmid: str) -> dict[str, Any] | None:
        data = self._request_json(
            f"/works/pmid:{urllib.parse.quote(str(pmid), safe='')}",
            {"select": RESOLUTION_SELECT},
        )
        return data if data.get("id") else None

    def search_work(self, title: Any, author_last_name: str | None = None) -> dict[str, Any] | None:
        title_text = text_or_none(title)
        if not title_text:
            return None
        data = self._request_json(
            "/works",
            {
                "search": title_text[:200],
                "per-page": 5,
                "select": RESOLUTION_SELECT,
            },
        )
        candidates = data.get("results") or []
        if not candidates:
            return None
        if author_last_name:
            needle = author_last_name.lower()
            for work in candidates:
                authorships = work.get("authorships") or []
                names = [
                    (authorship.get("author") or {}).get("display_name", "").lower()
                    for authorship in authorships
                ]
                if any(needle in name for name in names):
                    return work
        return candidates[0]

    def list_citers(
        self,
        parent_ids: list[str],
        *,
        cursor: str = "*",
        per_page: int = 100,
    ) -> OpenAlexPage:
        compact_ids = [compact_openalex_id(parent_id) for parent_id in parent_ids]
        compact_ids = [parent_id for parent_id in compact_ids if parent_id]
        if not compact_ids:
            return OpenAlexPage([], None, 0)
        if len(compact_ids) > 100:
            raise ValueError("OpenAlex OR filters support at most 100 values per request.")
        data = self._request_json(
            "/works",
            {
                "filter": "cites:" + "|".join(compact_ids),
                "per-page": max(1, min(100, per_page)),
                "cursor": cursor,
                "select": WORK_SELECT,
            },
        )
        meta = data.get("meta") or {}
        return OpenAlexPage(
            results=data.get("results") or [],
            next_cursor=meta.get("next_cursor"),
            count=meta.get("count"),
        )


def normalize_work(work: dict[str, Any]) -> dict[str, Any]:
    openalex_id = compact_openalex_id(work.get("id"))
    primary_location = work.get("primary_location") or {}
    source = primary_location.get("source") or {}
    primary_topic = work.get("primary_topic") or {}
    topic_domain = primary_topic.get("domain") or {}
    referenced = [compact_openalex_id(item) for item in (work.get("referenced_works") or [])]
    referenced = [item for item in referenced if item]
    return {
        "openalex_id": openalex_id,
        "doi": clean_doi(work.get("doi")),
        "title": work.get("display_name") or work.get("title"),
        "publication_date": work.get("publication_date"),
        "publication_year": work.get("publication_year"),
        "work_type": work.get("type") or work.get("type_crossref"),
        "is_retracted": work.get("is_retracted"),
        "cited_by_count": work.get("cited_by_count"),
        "source_id": compact_openalex_id(source.get("id")),
        "source_name": source.get("display_name"),
        "topic_id": compact_openalex_id(primary_topic.get("id")),
        "topic_name": primary_topic.get("display_name"),
        "topic_domain": topic_domain.get("display_name"),
        "referenced_works_json": json_dumps(referenced),
        "raw_json": json_dumps(work),
    }


def edges_from_work(work: dict[str, Any], parent_ids: set[str]) -> list[tuple[str, str]]:
    source_id = compact_openalex_id(work.get("id"))
    if not source_id:
        return []
    referenced = {compact_openalex_id(item) for item in (work.get("referenced_works") or [])}
    referenced.discard(None)
    parent_ids = {compact_openalex_id(item) for item in parent_ids}
    return [(source_id, target_id) for target_id in sorted(referenced.intersection(parent_ids))]


def openalex_url(openalex_id: str | None) -> str | None:
    return full_openalex_id(openalex_id)


def _openalex_doi_url(value: Any) -> str | None:
    doi = clean_doi(value)
    if not doi or "," in doi or "&" in doi:
        return None
    return doi_url(doi)


def _read_http_error_body(exc: urllib.error.HTTPError) -> str:
    try:
        body = exc.read()
    except Exception:
        return ""
    try:
        return body.decode("utf-8", errors="replace")
    except Exception:
        return ""


def _retry_after_seconds(
    retry_after: str | None,
    default_sleep: float,
    attempt: int,
) -> float:
    if retry_after:
        try:
            return min(300.0, max(1.0, float(retry_after)))
        except ValueError:
            pass
    return min(300.0, max(default_sleep, 10.0 * (2.0**attempt)))

File: R/openalex.R
#' @title OpenAlex API client
#' @description Port of the original Python `retraction_pollution.openalex`
#'   module. Provides an HTTP client for the OpenAlex REST API (`/works`,
#'   `/works/pmid:<pmid>`) plus helpers for normalizing raw work records into
#'   the storage schema and extracting citation edges from `referenced_works`.
#'
#'   The pure helpers (`normalize_work`, `edges_from_work`, `openalex_url`)
#'   depend only on the `util.R` cleaners (`clean_doi`, `compact_openalex_id`,
#'   `doi_url`, `full_openalex_id`, `json_dumps`, `text_or_none`).

library(R6)

OPENALEX_API <- "https://api.openalex.org"

WORK_SELECT <- paste(
  c("id", "doi", "display_name", "title", "publication_date",
    "publication_year", "type", "type_crossref", "is_retracted",
    "cited_by_count", "referenced_works", "primary_location",
    "primary_topic", "topics"),
  collapse = ","
)

RESOLUTION_SELECT <- paste(
  c("id", "doi", "display_name", "title", "publication_date",
    "publication_year", "type", "is_retracted", "cited_by_count",
    "referenced_works", "primary_location", "primary_topic", "topics",
    "authorships"),
  collapse = ","
)

# Null-coalescing helper (matches Python's `x or default`).
`%||%` <- function(a, b) if (is.null(a) || length(a) == 0L) b else a

#' Construct an `OpenAlexError` condition object (without raising).
#' @param msg character message.
#' @param status_code integerish HTTP status, or `NA`.
#' @param response_text character response body, or `NA`.
#' @return An object of class `c("OpenAlexError", "error", "condition")`.
new_openalex_error <- function(msg, status_code = NA_integer_,
                              response_text = NA_character_) {
  structure(
    list(message = msg, status_code = status_code,
         response_text = response_text),
    class = c("OpenAlexError", "error", "condition")
  )
}

#' Raise an `OpenAlexError`.
#' @inheritParams new_openalex_error
#' @return Never; signals an error of class `OpenAlexError`.
openalex_error <- function(msg, status_code = NA_integer_,
                           response_text = NA_character_) {
  stop(new_openalex_error(msg, status_code = status_code,
                         response_text = response_text))
}

#' Raise an `OpenAlexError`.
#' @noRd
stop_openalex <- function(msg, status_code = NA_integer_,
                          response_text = NA_character_) {
  openalex_error(msg, status_code = status_code,
                 response_text = response_text)
}

#' Construct an `OpenAlexRateLimitError` condition object (without raising).
#' @inheritParams new_openalex_error
#' @return An object of class
#'   `c("OpenAlexRateLimitError", "OpenAlexError", "error", "condition")`.
new_openalex_rate_limit_error <- function(msg, status_code = 429L,
                                          response_text = NA_character_) {
  structure(
    list(message = msg, status_code = status_code,
         response_text = response_text),
    class = c("OpenAlexRateLimitError", "OpenAlexError", "error", "condition")
  )
}

#' Raise an `OpenAlexRateLimitError`.
#' @inheritParams new_openalex_error
#' @return Never; signals an error of class `OpenAlexRateLimitError`.
openalex_rate_limit_error <- function(msg, status_code = 429L,
                                       response_text = NA_character_) {
  stop(new_openalex_rate_limit_error(msg, status_code = status_code,
                                     response_text = response_text))
}

#' Raise an `OpenAlexRateLimitError`.
#' @noRd
stop_openalex_rate_limit <- function(msg, status_code = 429L,
                                     response_text = NA_character_) {
  openalex_rate_limit_error(msg, status_code = status_code,
                            response_text = response_text)
}

#' Construct an `OpenAlexPage` data-carrier object.
#'
#' @param results list of raw work records.
#' @param next_cursor character cursor for the next page, or `NULL`/`NA`.
#' @param count integerish total result count, or `NA`.
#' @return An `OpenAlexPage` object.
OpenAlexPage <- function(results = list(), next_cursor = NULL,
                         count = NA_integer_) {
  structure(
    list(
      results = results,
      next_cursor = if (is_missing(next_cursor)) NULL else next_cursor,
      count = count
    ),
    class = "OpenAlexPage"
  )
}

#' @title OpenAlex HTTP client (R6)
#' @description Wraps `httr2` requests against the OpenAlex `/works` endpoint.
#'   Retries on 429 and 5xx, treats 404 as an empty result, and sleeps
#'   `request_delay` seconds before each request. The private `request_json`
#'   method is overridable by subclasses for testing.
#' @noRd
OpenAlexClient <- R6::R6Class(
  "OpenAlexClient",
  cloneable = FALSE,
  public = list(
    api_key = NULL,
    email = NULL,
    base_url = NULL,
    retries = NULL,
    request_delay = NULL,
    rate_limit_sleep = NULL,

    initialize = function(api_key = NA_character_, email = NA_character_,
                          base_url = OPENALEX_API, retries = 6L,
                          request_delay = 0.35, rate_limit_sleep = 60.0) {
      self$api_key <- if (is_missing(api_key)) NA_character_ else api_key
      self$email <- if (is_missing(email)) NA_character_ else email
      self$base_url <- sub("/+$", "", base_url)
      self$retries <- as.integer(retries)
      self$request_delay <- as.numeric(request_delay)
      self$rate_limit_sleep <- as.numeric(rate_limit_sleep)
      invisible(self)
    },

    get_works_by_dois = function(dois) {
      values <- unique(vapply(dois, function(d) {
        d2 <- clean_doi(d)
        if (is.na(d2) || grepl(",", d2, fixed = TRUE) ||
            grepl("&", d2, fixed = TRUE)) return(NA_character_)
        doi_url(d2)
      }, character(1)))
      values <- values[!is.na(values)]
      if (length(values) == 0L) return(list())
      if (length(values) > 100L) {
        out <- list()
        for (start in seq(1L, length(values), by = 100L)) {
          end <- min(start + 99L, length(values))
          out <- c(out, private$get_works_by_doi_values(values[start:end]))
        }
        out
      } else {
        private$get_works_by_doi_values(values)
      }
    },

    get_work_by_pmid = function(pmid) {
      pmid_text <- clean_pmid(pmid)
      if (is.na(pmid_text)) return(NULL)
      data <- private$request_json(
        paste0("/works/pmid:", utils::URLencode(pmid_text, reserved = TRUE)),
        list(select = RESOLUTION_SELECT)
      )
      if (is.list(data) && !is.null(data$id) && !is.na(data$id)) data else NULL
    },

    search_work = function(title, author_last_name = NULL) {
      title_text <- text_or_none(title)
      if (is.na(title_text)) return(NULL)
      data <- private$request_json("/works", list(
        search = substr(title_text, 1, 200),
        `per-page` = 5L,
        select = RESOLUTION_SELECT
      ))
      candidates <- data$results %||% list()
      if (length(candidates) == 0L) return(NULL)
      author <- text_or_none(author_last_name)
      if (!is.na(author)) {
        needle <- tolower(author)
        for (work in candidates) {
          authorships <- work$authorships %||% list()
          names <- vapply(authorships, function(a) {
            au <- a$author %||% list()
            tolower(au$display_name %||% "")
          }, character(1))
          if (any(stringr::str_detect(names, fixed(needle)))) return(work)
        }
      }
      candidates[[1]]
    },

    list_citers = function(parent_ids, cursor = "*", per_page = 100L) {
      compact_ids <- vapply(parent_ids, compact_openalex_id, character(1))
      compact_ids <- compact_ids[!is.na(compact_ids) & compact_ids != ""]
      if (length(compact_ids) == 0L) return(OpenAlexPage(list(), NULL, 0L))
      if (length(compact_ids) > 100L) {
        stop("OpenAlex OR filters support at most 100 values per request.",
             call. = FALSE)
      }
      data <- private$request_json("/works", list(
        filter = paste0("cites:", paste(compact_ids, collapse = "|")),
        `per-page` = max(1L, min(100L, as.integer(per_page))),
        cursor = cursor,
        select = WORK_SELECT
      ))
      meta <- data$meta %||% list()
      OpenAlexPage(
        results = data$results %||% list(),
        next_cursor = meta$next_cursor,
        count = meta$count %||% NA_integer_
      )
    },

    rate_limit_status = function() {
      private$request_json("/rate-limit")
    }
  ),

  private = list(
    request_json = function(path, params = list()) {
      if (self$request_delay > 0) Sys.sleep(self$request_delay)
      url <- paste0(self$base_url, path)
      query <- as.list(params)
      if (!is.na(self$api_key) && nchar(self$api_key) > 0) {
        query$api_key <- self$api_key
      }
      if (!is.na(self$email) && nchar(self$email) > 0) {
        query$mailto <- self$email
      }
      req <- httr2::request(url)
      if (length(query) > 0) req <- httr2::req_url_query(req, !!!query)
      req <- httr2::req_headers(
        req,
        Accept = "application/json",
        `User-Agent` = "RetractionPollution/0.1 (mailto optional; OpenAlex research)"
      )
      req <- httr2::req_retry(
        req,
        max_tries = self$retries,
        max_seconds = 60,
        is_transient = function(resp) {
          httr2::resp_status(resp) == 429 || httr2::resp_status(resp) >= 500
        }
      )
      req <- httr2::req_error(req, is_error = function(resp) FALSE)

      resp <- tryCatch(
        httr2::req_perform(req),
        httr2_error = function(e) {
          stop_openalex(paste("OpenAlex request failed:",
                              conditionMessage(e)))
        }
      )

      status <- httr2::resp_status(resp)
      if (status == 404) return(list())
      if (status >= 400) {
        body <- tryCatch(httr2::resp_body_string(resp),
                         error = function(e) "")
        if (is.null(body) || is.na(body)) body <- ""
        stop_openalex(
          paste0("OpenAlex HTTP ", status, " for ", path, ": ",
                 substr(body, 1, 300)),
          status_code = status,
          response_text = body
        )
      }
      httr2::resp_body_json(resp)
    },

    get_works_by_doi_values = function(values) {
      if (length(values) == 0L) return(list())
      tryCatch({
        data <- private$request_json("/works", list(
          filter = paste0("doi:", paste(values, collapse = "|")),
          `per-page` = max(1L, min(100L, length(values))),
          select = RESOLUTION_SELECT
        ))
        data$results %||% list()
      }, OpenAlexError = function(e) {
        if (is.na(e$status_code) || e$status_code != 400) stop(e)
        if (length(values) == 1L) return(list())
        mid <- length(values) %/% 2L
        c(private$get_works_by_doi_values(values[seq_len(mid)]),
          private$get_works_by_doi_values(values[(mid + 1L):length(values)]))
      })
    }
  )
)

#' Normalize a raw OpenAlex work record into the storage schema.
#'
#' @param work list, a raw OpenAlex work record.
#' @return named list with `WORK_COLUMNS`.
normalize_work <- function(work) {
  if (is.null(work)) work <- list()
  openalex_id <- compact_openalex_id(work$id)
  primary_location <- work$primary_location %||% list()
  source <- primary_location$source %||% list()
  primary_topic <- work$primary_topic %||% list()
  topic_domain <- primary_topic$domain %||% list()
  referenced <- vapply(work$referenced_works %||% list(),
                      compact_openalex_id, character(1))
  referenced <- referenced[!is.na(referenced) & referenced != ""]
  list(
    openalex_id = openalex_id,
    doi = clean_doi(work$doi),
    title = work$display_name %||% work$title,
    publication_date = work$publication_date,
    publication_year = work$publication_year,
    work_type = work$type %||% work$type_crossref,
    is_retracted = work$is_retracted,
    cited_by_count = work$cited_by_count,
    source_id = compact_openalex_id(source$id),
    source_name = source$display_name %||% NA_character_,
    topic_id = compact_openalex_id(primary_topic$id),
    topic_name = primary_topic$display_name %||% NA_character_,
    topic_domain = topic_domain$display_name %||% NA_character_,
    referenced_works_json = json_dumps(referenced),
    raw_json = json_dumps(work)
  )
}

#' Extract citation edges from a work's `referenced_works`, intersected
#' with the set of parent IDs we are currently crawling from.
#'
#' @param work list, a raw OpenAlex work record.
#' @param parent_ids character vector of compact parent OpenAlex IDs.
#' @return list of `c(source, target)` character pairs.
edges_from_work <- function(work, parent_ids) {
  source_id <- compact_openalex_id(work$id)
  if (is.na(source_id) || source_id == "") return(list())
  referenced <- vapply(work$referenced_works %||% list(),
                     compact_openalex_id, character(1))
  referenced <- referenced[!is.na(referenced) & referenced != ""]
  parents <- vapply(parent_ids, compact_openalex_id, character(1))
  parents <- parents[!is.na(parents) & parents != ""]
  hits <- intersect(referenced, parents)
  if (length(hits) == 0L) return(list())
  hits <- sort(unique(hits))
  lapply(hits, function(t) c(source_id, t))
}

#' Build the full OpenAlex URL for a compact ID.
#' @param openalex_id character compact OpenAlex ID or `NA`.
#' @return full URL or `NA_character_`.
openalex_url <- function(openalex_id) {
  full_openalex_id(openalex_id)
}